node/src/reactor/joiner.rs

```diff
@@ -586,13 +602,15 @@ impl reactor::Reactor for Reactor {
         let block_by_height_fetcher = Fetcher::new("block_by_height", config.fetcher, registry)?;
 
         let block_by_hash_fetcher = Fetcher::new("block", config.fetcher, registry)?;
-        let trie_fetcher = Fetcher::new("trie", config.fetcher, registry)?;
         let block_header_and_finality_signatures_by_height_fetcher =
             Fetcher::new("block_header_by_height", config.fetcher, registry)?;
 
         let block_header_by_hash_fetcher: Fetcher<BlockHeader> =
             Fetcher::new("block_header", config.fetcher, registry)?;
 
+        let trie_or_chunk_fetcher = Fetcher::new("trie_or_chunk", config.fetcher, registry)?;
+        let trie_fetcher = TrieFetcher::new();
```

Comment:
If we pass in `Fetcher<TrieOrCunk>` into `TrieFetcher`'s constructor then we can get away with just single struct and two (instead of four) events/requests. Also, it would be conceptually simpler if `TrieFetcher` "built" on top of generic `Fetcher`.

Reply:
Ah, yes. I was afraid of it. Unfortunately, it makes sense we cannot do it.